### ISSUE TITLE
Mixed-type applicative

### ISSUE THREAD

**Issue by bepremeg** [gcanti/fp-ts#315]
Not sure how best to phrase this but in a few instances I have functions such this:

```Typescript
type Interpreter<T> = (obj: Object) => validation.Validation<string, T>;

function map3<A1, A2, A3, Value>(
  f: (a1: A1, a2: A2, a3: A3) => Value,
  interpreter1: Interpreter<A1>,
  interpreter2: Interpreter<A2>,
  interpreter3: Interpreter<A3>
): Interpreter<Value> {
  return (json: Object) => 
    interpreter1(json)
      .chain(a1 => interpreter2(json)
         .chain(a2 => interpreter3(json)
           .map(a3 => f(a1, a2, a3))
         )
      );
}
```

This works, but is very repetitive. Especially, when going up to `map8`. It is also very close to validation.traverse, except for the mixed types (`A1`, `A2` iso a single `A`). Is this something that can be achieved by fp-ts in a more concise way?

Thanks

**Comment by gcanti** [comment]
@bepremeg that's the most concise way to do it IMO. Theoretically `map2`, `map3`, ..., `mapN` can be generalized and derived from `Apply` 

```ts
import { HKT } from 'fp-ts/lib/HKT'
import { Apply } from 'fp-ts/lib/Apply'

function map2<F>(F: Apply<F>): <A, B, R>(fa: HKT<F, A>, fb: HKT<F, B>, f: (a: A, b: B) => R) => HKT<F, R> {
  return <A, B, R>(fa: HKT<F, A>, fb: HKT<F, B>, f: (a: A, b: B) => R) => F.ap(F.map(fa, a => (b: B) => f(a, b)), fb)
}

function map3<F>(
  F: Apply<F>
): <A, B, C, R>(fa: HKT<F, A>, fb: HKT<F, B>, fc: HKT<F, C>, f: (a: A, b: B, c: C) => R) => HKT<F, R> {
  return <A, B, C, R>(fa: HKT<F, A>, fb: HKT<F, B>, fc: HKT<F, C>, f: (a: A, b: B, c: C) => R) =>
    F.ap(F.ap(F.map(fa, a => (b: B) => (c: C) => f(a, b, c)), fb), fc)
}

// etc...
```

but you end up writing more code as you must define an instance of `Apply` for `Interpreter`

**Comment by bepremeg** [comment]
Thanks a lot. I had also found a way to use applicative to do what I wanted, but just as in your example, Typescript's lack of infix notation makes it kind of hard to follow what's going on as N increases. Still interesting to see your more general approach than mine.

In the end I have come up with a `mapRecord` function that achieves what I need with even more ease and type safety for the user.

```typescript
type InterpreterOptionalRecord<A> = { readonly [P in keyof A]: Interpreter<Option<A[P]>> };

function interpretRecord<A>(record: InterpreterOptionalRecord<A>): Interpreter<A> {
  return (json: Object) => {
    const result = {} as Partial<A>; // Because we want to be able to set every property as undefined 
    const validations = new Array<Validation<any>>();
    // tslint:disable-next-line:forin
    for (const k in record) {
      const validationOutcome: Validation<Option<A[keyof A]>> = record[k](json);
      // set all results where validation is ok.
      validationOutcome.map(maybeValue => (result[k] = maybeValue.toUndefined())); // forEach
      // combine all fails, map away the ok's (They all have a different type)
      validations.push(validationOutcome.map(_ => {}));
    }
    // The sequenced validation is ok iff all sub validations are ok.
    // we already computed the "real" result in the for loop above
    return sequence(validations).map(_ => result as A); 
  };
}

function sequence<T>(validations: Array<Validation<T>>) {
  return traversable.sequence(validation, array)(validations);
}
```

I can't say I'm very proud of the way the `result` record is created as a side effect in the imperative for-loop, but at least to the user of the function every is functional. 

e.g.

```typescript
const circle: Interpreter<ol.style.Circle> = st.mapRecord(olc.Circle, {
  radius: st.reqField("radius", st.num),
  fill: st.optField("fill", fill),
  stroke: st.optField("stroke", stroke),
  snapToPixel: st.optField("snapToPixel", st.bool)
});
```

where

```typescript
function Circle(circleOptions: {
  radius: number; //
  fill?: ol.style.Fill;
  stroke?: ol.style.Stroke;
  snapToPixel?: boolean;
}): ol.style.Circle {
  return new ol.style.Circle(circleOptions);
}
```

Now that I look at the code again, I should be able to optimize using a traverse for combining the failures and setting the record properties as a side-effect, but sometimes you have to know when to stop and just deliver a feature to the customer I guess.
